### ISSUE TITLE
ipyparallel - exception handling in clients?

### ISSUE THREAD

**Issue by isaacovercast** [dereneaton/ipyrad#49]
Is it possible to raise an exception from an ipyclient that'd get back to the calling process in a sensible way? Right now if a call fails in an ipyclient i can call sys.exit() or raise() and it doesn't do what i'd really want, which is to bail out an notify the user. You have more exp w/ it so maybe there's a good answer?

**Comment by dereneaton** [comment]
The `_clientwrapper()` should be catching any errors that happen within the clients since the step functions are run inside a `try` statement and the wrapper has a bunch of `except` calls. It will print the errors to screen differently depending on how the error is handled by the wrapper. It seems to be working well for me, to the point I hardly look at the debugger any more. I just pushed some changes that make it more clear, but I'll also describe below.

Example 1:
[ipyclient]
 `raise SystemExit("I'm outta here")`
[parent] 
```
...(traceback)  
SystemExit: I'm outta here
```
This error is caught by the wrapper because we have `except SystemExit as inst` in the code. This catches the error and saves "I'm outta here" in the var `inst`. We can then print inst to screen and keep going on our way, or we can call `raise` to reraise the error in which case it will interrupt the job and print the error to the screen with a traceback. For Systemexit we use raise.

----------------------

Example 2: 
[ipyclient] 
`assert x>3, "report warning"`
[parent]
```
...(traceback)
IPyradError: report warning
```
For AssertionError I catch the error and raise it again as an IPyradError. Not too much point in doing this, but it makes it look like we know what we're doing if we handle the error ourselves.

-----------------------


The way I have it now:
`KeyboardInterrupt` logs to tell us why the job stopped, and then is reraised
`SystemExit` logs to tell us the system stopped our job, then is reraised
`IPyradError` prints an error to screen but does not raise 
`ipp.NoEnginesRegistered` prints directions to screen to get ipcluster running, raise as IPyradError
`AssertionError` is caught along with `IPyradError` and is reraised as an `IPyradError`.

**Comment by dereneaton** [comment]
tldr; `raise` after handling the exception is the important component of whether or not the program bails. If you want it to not bail but still save the error for you to look at, then pass it to a var like `inst`.

**Comment by isaacovercast** [comment]
got it. makes good sense, thx.

**Comment by dereneaton** [comment]
I'm cleaning up the code now since there is a lot of internal exception handling junk leftover that is now handled by the wrapper.
